File: src/speed_dating.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	file, err := os.Open("Divvy_Trips_20250501.csv")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var minDate, maxDate string
	first := true

	// Skip the header
	if scanner.Scan() {
		// do nothing, skip first line
	}

	for scanner.Scan() {
		line := scanner.Text()
		fields := strings.SplitN(line, ",", 4)
		if len(fields) < 3 {
			continue
		}

		// Get just the mm/dd/yyyy portion (first 10 chars) from field 2 and 3
		start := fields[1]
		stop := fields[2]
		if len(start) < 10 || len(stop) < 10 {
			continue
		}
		startDate := reformatDate(start[:10]) // mm/dd/yyyy -> yyyymmdd
		stopDate := reformatDate(stop[:10])

		if first {
			minDate, maxDate = startDate, stopDate
			first = false
		} else {
			if startDate < minDate {
				minDate = startDate
			}
			if stopDate > maxDate {
				maxDate = stopDate
			}
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Scanner error:", err)
		return
	}

	fmt.Printf("Earliest: %s/%s/%s\n", minDate[4:6], minDate[6:], minDate[:4])
	fmt.Printf("Latest:   %s/%s/%s\n", maxDate[4:6], maxDate[6:], maxDate[:4])
}

func reformatDate(d string) string {
	// Assumes mm/dd/yyyy
	// Returns yyyymmdd as a string
	return d[6:10] + d[0:2] + d[3:5]
}

File: src/speed_dating.py
import csv

def extract_min_max_dates(filename):
    min_date = None
    max_date = None

    with open(filename, newline='') as f:
        reader = csv.reader(f)
        next(reader)  # skip header
        for row in reader:
            for field_index in [1, 2]:  # start time, end time
                date = row[field_index].split(' ')[0]  # mm/dd/yyyy
                #print("date",date)
                m, d, y = date.split('/')
                compact = f"{y}{m}{d}"
                if compact < min_date or not min_date:
                    min_date = compact
                if compact > max_date or not max_date:
                    max_date = compact

    print("Earliest:", f"{min_date[4:6]}/{min_date[6:]}/{min_date[:4]}")
    print("Latest:  ", f"{max_date[4:6]}/{max_date[6:]}/{max_date[:4]}")

# Usage
extract_min_max_dates("Divvy_Trips_20250501.csv")

